📖 Document e2e tests

lentzi90 opened the pull request:
**What this PR does / why we need it**:

Add some docs on e2e tests, especially upgrade BMO/Ironic

lentzi90 commented:
/test-v1b1-ubuntu-integration

lentzi90 commented:
/test-v1b1-ubuntu-integration

furkatgofurov7 commented on docs/e2e-test.md:
```diff
@@ -26,3 +26,31 @@ After a finished test, cleanup is performed automatically. If the test setup fai
+
+### Upgrade tests
+
+Both Ironic and BMO upgrade tests currently only check that a new version can be rolled out (by going from `latest` to `main`).
+However, they do not actually upgrade from some older release to a newer since we do not yet have the full upgrade test in place for CAPM3.
+Once this is in place, the idea is to upgrade CAPM3 from the previous minor release to the latest.
+Ironic and BMO will then also be upgraded in similar manner using versions that are compatible with CAPM3.
```
I would suggest not to merge this as is but rather wait for the proper upgrade to land (consider #466 will be in soon) and then update these notes if there is no real rush on merging it.

lentzi90 replied:
Sure we can wait with this. #466 is now merged, but I guess it'll be another few weeks until we have refactored the upgrade to include BMO and Ironic. It is fine for me to add the documentation after that but just wanted to explain that it will probably take a while until this part is completed.

furkatgofurov7 commented on docs/e2e-test.md:
```diff
@@ -26,3 +26,31 @@ After a finished test, cleanup is performed automatically. If the test setup fai
+1. Node reuse
+1. Re-pivoting
+
+*) Inspection is actually run [in the middle of the remediation test](https://github.com/metal3-io/cluster-api-provider-metal3/blob/8d08f375de93a793f839b42b5ec40e6bebf98664/test/e2e/remediation_test.go#L108) for practical reasons at the moment.
```
This part with * looks a bit weird when viewed this file with "Display the rich diff"
![image](https://user-images.githubusercontent.com/40443040/157208740-565a8546-7c2b-4d7a-aa6d-d1503240d3d6.png)

lentzi90 replied:
Not sure I understand. I added a `*` after inspection since it is a bit special that it runs in the middle of the remediation test, and added the explanation here. Any suggestion how to make it clearer?

furkatgofurov7 replied:
@lentzi90 yes, exactly that is what I meant, sorry was not clear, how about instead of **¹)**  just **¹** ? i.e:

¹Explanation here

lentzi90 replied:
Updated!

kashifest commented:
/test-v1b1-ubuntu-integration

kashifest commented:
/test-v1b1-ubuntu-integration

lentzi90 commented:
Also added trigger phrase for upgrade tests in the readme now and updated the docs to reflect that the upgrade tests are merged.
Please note that BMO and Ironic upgrades are not included in the e2e upgrade test. They are still part of the "normal" e2e test.

lentzi90 commented:
/test-v1b1-ubuntu-integration

furkatgofurov7 commented on docs/e2e-test.md:
```diff
@@ -26,3 +26,30 @@ After a finished test, cleanup is performed automatically. If the test setup fai
+After this, the controllers are moved to the target cluster during pivoting.
+From this point on all tests are run in the target cluster, until the controllers are moved back to the ephemeral cluster during re-pivoting.
+
+### Upgrade tests
+
+Both Ironic and BMO upgrade tests currently only check that a new version can be rolled out (by going from `latest` to `main`).
+However, they do not actually upgrade from some older release to a newer since they are not yet integrated with the full upgrade test in CAPM3.
+Once this is in place, the idea is to upgrade Ironic and BMO together with CAPM3 from the previous minor release to the latest.
```
As discussed offline, I would suggest removing **### Upgrade tests** sections since it might confuse others thinking that those tests are run when e2e upgrade tests are triggered since you are adding e2e upgrade trigger phrases under `Upgrade tests` in the root readme file. 
Instead, we could:

- add Ironic and BMO upgrade related just after feature tests description on line 50 for now since they are part of the  `normal e2e tests`
- once we get Ironic and BMO tests out of `normal e2e tests`  and include them in the `e2e upgrade tests`, we can put those notes + management cluster upgrade related notes back again here.

lentzi90 replied:
Updated, please take a look!

furkatgofurov7 replied:
Looks good, thanks!

furkatgofurov7 reviewed (approved):
/approve

furkatgofurov7 commented:
This should skip running whole integration tests after https://github.com/metal3-io/project-infra/pull/305/commits/737bb314ee736fba4dc49fd7581751d1a25fd043, let's give it a try :)

/test-v1b1-ubuntu-integration

furkatgofurov7 commented:
/test-v1b1-ubuntu-integration

furkatgofurov7 commented:
> This should skip running whole integration tests after [metal3-io/project-infra@737bb31](https://github.com/metal3-io/project-infra/commit/737bb314ee736fba4dc49fd7581751d1a25fd043), let's give it a try :)

Looks like working now after followup https://github.com/metal3-io/project-infra/pull/362/commits/34dc22658c765f35e6ac454939f80ef47c12498d

kashifest commented:
/lgtm
Can you change the title ? It looks a bit odd.
/hold

lentzi90 commented:
Done :heavy_check_mark:

kashifest commented:
/hold cancel
